README
Astro::Coords
=============

This module provides an object oriented approach to astronomical
coordinate transformations. It can be used to convert between
J2000, B1950, Galactic and super Galactic coordinates.  It can
calculate the azimuth and elevation of a source specified in
equatorial coordinates, orbital elements or by name (if a planet).
It can also determine whether the specified coordinates are observable
for the given telescope and determine rise and set times. Doppler
corrections can also be calculated.

This software is still a beta release but is in use in a production
environment. The interfaces may still change slightly.

It is a wrapper class around Astro::SLA that simplifies the handling of
astronomical coordinate systems. 

Requirements
------------

This class uses Astro::SLA, Astro::Telescope as well as
DateTime and Time::Piece.

This release begins the migration from Time::Piece to DateTime
for specification of epochs. Both are supported at present but
at some point Time::Piece support may be removed.

Building
--------

  perl Build.PL
  ./Build
  ./Build test
  ./Build install


Author
------

This module was written by Tim Jenness <tjenness@cpan.org>
of the Joint Astronomy Centre.

Copyright (C) 2008,2010 Science and Technologiy Facilities Council.
Copyright (C) 2001-2005 Particle Physics and Astronomy Research Council.
All Rights Reserved.

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful,but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program; if not, write to the Free Software Foundation, Inc., 59 Temple
Place,Suite 330, Boston, MA  02111-1307, USA

TODO
----

  + Add support for JPL ephemeris when calculating planet positions
  + Add support for Horizons web service when generating positions
    of planetary moons

Changes
-------

 v0.01 - initial test release to CPAN

 v0.02 - Add Astro::Coords::Calibration and Astro::Coords::Fixed

 v0.03 - Add isObservable method
	 Spun off Astro::Telescope into its own CPAN distribution

 v0.04 - Internal release

 v0.05 - - Check format of input arguments (for sexagesimal parsing)
         + Add rise, set and transit time calculation
         + Add Interpolated RA/Dec [Astro::Coords::Interpolated]
         + Add calculate method
         - Fix orbital elements [requires recent slalib]
         + Can now ask for J2000 for all coordinate types
         + Add name and comment method for tagging a coordinate object

 v0.07 - Add support for parallax and proper motion, epoch and equinox
         + ra2000 and dec2000 methods always return RA/Dec J2000 for epoch 2000.0
          (equatorial only)

 v0.10 - + Major reorganization internally to use (the new) Astro::Coords::Angle
           objects for all angles. Should simplify formatting of return values.
         + datetime() method now supports DateTime objects for time
           specification. Time::Piece is now deprecated.
         + Add radial velocity support
         + Add doppler correction (experimental)
         + New method to obtain the native Equatorial encoding that was used
         + set_time and rise_time now works properly for the Moon (and Sun)
         + rise_time can now be configured to return the nearest rise time
           not just the next rise_time.
         + planet diameter is now available
         + Galactic and ecliptic conversions now in base class (moved from Equatorial)
         + radec1950 method available
         + Switch to GPL license

v0.12 - + Update Elements interface
        + Improve performance of DateTime usage (caching TimeZone object)
        + Support barycentric velocities
        + Astro::Coords::Offset now has invert(), clone() and posang() method
        + Astro::Coords::Angle now has clone() and negate() methods.
        + Astro::Coords::Offset now allows more systems.
        + Fix problem with fractions seconds in LST calculation
        + Fix problem with calculate() method trashing the cached DateTime.
        + Uses Module::Build


ChangeLog
2010-12-02  Tim Jenness  <tjenness@cpan.org>

	* Build.PL: Use Module::Build

2008-12-30  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords.pm: Fix some warnings.

	* lib/Astro/Coords/Angle/Hour.pm: Fix some warnings.

2008-06-02  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords/Offset.pm: Fix regexps in system matching.

	* lib/Astro/Coords/Offset.pm: Ensure that the output system is a
	PTCS-compliant system.

2008-04-09  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords.pm (_lst): Fix problems with fractional seconds in LST
	calculation and in datetime cloning.

2007-05-30  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords.pm (calculate): Make sure that the calculate()
	method does not trash the existing datetime().

2007-02-05  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords/Angle/Hour.pm: Fix constructor so that it
	correctly recognizes decimal hours.

2006-10-18  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords/Offset.pm (system): Allow B1900 and J2006.4 etc
	as offset system.

2006-01-09  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords/Offset.pm (new): Allow for case-insensitive
	keys in constructor.  Allow "Az/El" to be a synonym for AZEL.

2005-12-08  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords/Offset.pm (invert): Add invert method

	* lib/Astro/Coords/Angle.pm (negate): Add negate method

	* lib/Astro/Coords/Offset.pm (clone): Add clone method

	* lib/Astro/Coords/Angle.pm (clone): Add clone method

	* lib/Astro/Coords/Offset.pm: Add DIRECT offset projection to match PTCS

2005-12-05  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords/Offset.pm (posang): Add posang() method.

2005-10-28  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords.pm (datetime_is_unsafe): Implement caching
	scheme for static dates. datetime_is_unsafe() method is required
	if the datetime object is to be tweaked externally. Large speed
	improvements.

2005-10-21  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords.pm (vbary): Add support for Barycentric velocities.

2005-08-24  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords/Equatorial.pm (set_vel_pars): Add ability to
	set the velocity parameters after object construction (but not in
	base class)

2005-08-06  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords/Fixed.pm: Override default meridian_time and
	transit_el calculations for fixed sources.

	* lib/Astro/Coords.pm (status): Allow meridian time and transit
	elevation to be undef.

2005-07-30  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords.pm (calculate): Fix missing DateTime check in
	the "calculate" method.

2005-07-08  Brad Cavanagh  <b.cavanagh@jach.hawaii.edu>

	* lib/Astro/Coords.pm (distance): Factor out RA/Dec lookup when
	calculating distance to reduce number of calls to radec() by 50%.

2005-05-20  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords.pm: Pre-create a DateTime::TimeZone object
	rather than using a 'UTC' string. This improves performance when
	using many calls to the datetime method.

2005-04-06  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords/Elements.pm (apparent): Include the target name
	in the error message when failing to determine the apparent RA/Dec.

2005-03-16  Tim Jenness  <tjenness@cpan.org>

	* t/basic.t: Test the array Elements constructor

	* lib/Astro/Coords.pm (new): Document upgrade to elements
	interface to accept array() and enhance the test for that case.

	* lib/Astro/Coords/Elements.pm (new): Now allow the return value
	from the array() method to be provided to the constructor.

	* lib/Astro/Coords/Elements.pm (array): For comets, now return the
	epoch of perihelion as the final member of the array and the epoch
	of elements as the initial element in the array. This allows both
	epochs to be retained without confusion (since the penultimate
	member will be undefined for the cometary case) but constitutes a
	change in the published interface.

	* lib/Astro/Coords/Elements.pm (apparent): Include the target name
	in the error message when failing to perturb the elements.

2005-02-21  Tim Jenness  <tjenness@cpan.org>

	---- RELEASE V0.10 to CPAN ----

2005-02-18  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords.pm: Rewrite the rise/set iterative code to
	handle all cases properly. Add new parameter "event" to control
	whether the nearest, previous or next rise/set time should be
	returned.

2005-02-16  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords/Angle.pm (_guess_units): Use Scalar::Util to
	determine whether the supplied string is a number. Now requires
	Scalar::Util v1.13

2005-02-14  Brad Cavanagh <b.cavanagh@jach.hawaii.edu>

	* lib/Astro/Coords/Angle.pm, lib/Astro/Coords/Angle/Hour.pm:
	Change default delimiter to the colon character to maintain
	backwards compatibility with v0.05.

2005-02-04  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords.pm (_iterative_el): Fix case when moon almost
	sets but doesn't quite make it (thanks for bug report from Beat
	Vontobel)

2005-01-05  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords/Equatorial.pm: Can not use vdiff in the radec
	method since vdiff requires the radec.

2004-12-09  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords.pm: Add preliminary support for radial
	velocities and velocity frames. Add doppler correction
	calculation.

2004-12-08  Tim Jenness  <tjenness@cpan.org>

	* t/coco.t: Finish test comparison with coco for simple B1950
	coordinate

	* lib/Astro/Coords.pm (native): Add native() method to keep track
	of original form.

2004-12-07  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords.pm: Use Astro::Coords::Angle[::Hour] for all
	return values that return angles. Make public methods for
	returning pairs of coordinates (azel, radec,apparent).

	* lib/Astro/Coords.pm: Add radec1950 method

2004-12-04  Tim Jenness  <tjenness@cpan.org>

	* t/angle.t: Add tests for Angle classes

2004-11-21  Tim Jenness  <tjenness@cpan.org>

	* lib/Astro/Coords/Planet.pm: Add diam() method to return the

lib/Astro/Coords/Fixed.pm
package Astro::Coords::Fixed;

=head1 NAME

Astro::Coords::Fixed - Manipulate coordinates that are fixed on the sky

=head1 SYNOPSIS

  $c = new Astro::Coords::Fixed( az => 180,
                                 el => 45,
				 units => 'degrees');

  $c = new Astro::Coords::Fixed( ha => '02:30:00.0',
				 dec => '45:30:03',
				 units => 'sexagesimal',
				 tel => $telescope,
			       );

=head1 DESCRIPTION

This subclass of C<Astro::Coords> allows for the manipulation
of coordinates that are fixed on the sky. Sometimes a telescope
should be commanded to go to a fixed location (eg for a calibration)
and this class puts those coordinates (Azimuth and elevation for telescopes
such as JCMT and Gemini and Hour Angle and Declination for equatorial
telescopes such as UKIRT) on the same footing as astronomical coordinates.

Note that Azimuth and elevation do not require the telescope latitude
whereas Hour Angle and declination does.

=cut


use 5.006;
use strict;
use warnings;
use Carp;

our $VERSION = '0.03';

use Astro::SLA ();
use Astro::Coords::Angle;
use base qw/ Astro::Coords /;

use overload '""' => "stringify";

=head1 METHODS

=head2 Constructor

=over 4

=item B<new>

Constructor. Recognizes hash keys "ha", "dec" and "az", "el".

  $c = new Astro::Coords::Fixed( az => 35, el => 30 );

  $c = new Astro::Coords::Fixed( ha => $ha, dec => $dec, tel => $tel);

Usually called via C<Astro::Coords> rather than directly.

Note that the declination is equivalent to "Apparent Dec" used
elsewhere in these classes.

Azimuth and Elevation is the internal format. Currently there is no
caching (so there is always overhead converting to apparent 
RA and Dec) since there is no cache flushing when the telescope
is changed.

A telescope is required (in the form of an C<Astro::Telescope> object)
if the position is specified as HA/Dec.

A name can be associated with this position.

=cut

sub new {
  my $proto = shift;
  my $class = ref($proto) || $proto;

  my %args = @_;

  # We will always calculate ha, dec, az and el
  my ($az, $el);

  # Create a new object
  my $c = bless { }, $class;

  # Store the telescope if we have one
  $c->telescope( $args{tel} ) if exists $args{tel};

  if (exists $args{ha} && exists $args{dec} and exists $args{tel}
     and UNIVERSAL::isa($args{tel}, "Astro::Telescope")) {
    # HA and Dec

    # Convert input args to radians
    my $ha = Astro::Coords::Angle::Hour->to_radians($args{ha},$args{units});
    my $dec = Astro::Coords::Angle->to_radians($args{dec}, $args{units});

    # Convert to "native" format
    my $lat = $args{tel}->lat;
    Astro::SLA::slaDe2h( $ha, $dec, $lat, $az, $el);

    $az = new Astro::Coords::Angle( $az, units => 'rad', range => '2PI');
    $el = new Astro::Coords::Angle( $el, units => 'rad');

    # native form
    $c->native( 'hadec' );

  } elsif (exists $args{az} and exists $args{el}) {
    # Az and El

    # Convert input args to radians
    $az = new Astro::Coords::Angle( $args{az}, units => $args{units},
				    range => '2PI' );
    $el = new Astro::Coords::Angle( $args{el}, units => $args{units});

    # native form
    $c->native( 'azel' );

  } else {
    return undef;
  }

  # Store the name
  $c->name( $args{name} ) if exists $args{name};

  # Store it in the object
  $c->azel( $az, $el );

  return $c;
}


=back

=head2 Accessor Methods

=over 4

=item B<azel>

Return azimuth and elevation (as two C<Astro::Coords::Angle> objects);

 ($az, $el) = $c->azel;

Can also be used to store the azimuth and elevation
(as C<Astro::Coords::Angle> objects)

  $c->_azel( $az, $el);

=cut

sub azel {
  my $self = shift;
  if (@_) {
    my ($az, $el) = @_;
    croak "Azimuth not an Astro::Coords::Angle object"
      unless UNIVERSAL::isa( $az, "Astro::Coords::Angle");
    croak "Elevation not an Astro::Coords::Angle object"
      unless UNIVERSAL::isa( $el, "Astro::Coords::Angle");
    $self->{Az} = $az;
    $self->{El} = $el;
  }
  return ($self->{Az}, $self->{El});
}

=back

=head2 General Methods

=over 4

=item B<type>

Returns the generic type associated with the coordinate system.
For this class the answer is always "FIXED".

This is used to aid construction of summary tables when using
mixed coordinates.

=cut

sub type {
  return "FIXED";
}

=item B<stringify>

Returns a string representation of the object. Returns
Azimth and Elevation in degrees.

=cut

sub stringify {
  my $self = shift;
  my $az = $self->az( format => "degrees" );
  my $el = $self->el( format => "degrees" );
  return "$az $el";
}

=item B<summary>

Return a one line summary of the coordinates.
In the future will accept arguments to control output.

  $summary = $c->summary();

=cut

sub summary {
  my $self = shift;
  my $name = $self->name;
  $name = '' unless defined $name;
  return sprintf("%-16s  %-12s  %-13s   AZEL",$name,
		 $self->az(format=>"s"),
		 $self->el(format =>"s"));
}

=item B<array>

Array summarizing the object. Retuns 
Return back 11 element array with first 3 elements being the
coordinate type (FIXED) and the az/el coordinates
(radians).

This method returns a standardised set of elements across all
types of coordinates.

=cut

sub array {
  my $self = shift;
  return ( $self->type, $self->az, $self->el,
	   undef, undef, undef, undef, undef, undef, undef, undef);
}

=item B<ha>

Get the hour angle for the currently stored LST. By default
returns format as for other angular methods.

  $ha = $c->ha;
  $ha = $c->ha( format => "deg" );

=cut

sub ha {
  my $self = shift;
  my %opt = @_;
  my $ha = ($self->hadec)[0];
  return $ha->in_format( $opt{format} );
}

=item B<apparent>

Return the apparent RA and Dec (as two C<Astro::Coords::Angle> objects)
for the current time [note that the apparent declination
is fixed and the apparent RA changes].

If no telescope is present the equator is used.

=cut

sub apparent {
  my $self = shift;

  my ($ra_app, $dec_app) = $self->_cache_read( "RA_APP", "DEC_APP" );

  if (!defined $ra_app || !defined $dec_app) {

    (my $ha, $dec_app) = $self->hadec;
    $ra_app = $self->_lst - $ha->radians;
    $ra_app = new Astro::Coords::Angle::Hour( $ra_app, units => 'rad', range => '2PI' );

    # should not cache the DEC_APP since we did not calculate it
    $self->_cache_write( "RA_APP" => $ra_app );
  }

  return( $ra_app, $dec_app);
}

=item B<hadec>

Return the Hour angle and apparent declination (as two C<Astro::Coords::Angle> objects).
If no telescope is present the equator is used.

 ($ha, $dec) = $c->hadec;

=cut

sub hadec {
  my $self = shift;

  my ($ha, $dec_app) = $self->_cache_read( "HA", "DEC_APP" );

  if (!defined $ha || !defined $dec_app) {

    my ($az, $el) = $self->azel;
    my $tel = $self->telescope;
    my $lat = ( defined $tel ? $tel->lat : 0.0);

    # First need to get the hour angle and declination from the Az and El
    Astro::SLA::slaDh2e($az->radians, $el->radians, $lat, $ha, $dec_app);

    $ha = new Astro::Coords::Angle::Hour( $ha, units => 'rad', range => 'PI');
    $dec_app = new Astro::Coords::Angle( $dec_app, units => 'rad');

    $self->_cache_write( "HA" => $ha, "DEC_APP" => $dec_app );
  }

  return ($ha, $dec_app);
}

=item B<ha_set>

For a fixed source, the setting Hour Angle has no meaning.

=cut

sub ha_set {
  return ();
}

=item B<meridian_time>

Meridian time is not defined for a fixed source.

=cut

sub meridian_time {
  return ();
}

=item B<transit_el>

Transit elevation is not defined for a fixed source. Always returns undef.

=cut

sub transit_el {
  return ();
}

=back

=head1 NOTES

Usually called via C<Astro::Coords>.

=head1 REQUIREMENTS

C<Astro::SLA> is used for all internal astrometric calculations.

=head1 AUTHOR

Tim Jenness E<lt>tjenness@cpan.orgE<gt>

=head1 COPYRIGHT

Copyright (C) 2001-2005 Particle Physics and Astronomy Research Council.
All Rights Reserved.

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful,but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program; if not, write to the Free Software Foundation, Inc., 59 Temple
Place,Suite 330, Boston, MA  02111-1307, USA

=cut

1;
